### PR TITLE
Fixes/dont cover taskbar with system decorations hidden

### PR DESCRIPTION
**Pull request by danwalmsley**

### CODE REVIEW

**Comment by danwalmsley** [comment]
Normal Size:
https://files.gitter.im/VitalElement/AvalonStudio/DnnH/image.png

Maximized:
https://files.gitter.im/VitalElement/AvalonStudio/zMCA/image.png

**Comment by danwalmsley** [comment]
This only actually affects Windows backend, as on mac and Linux windows by default in this condition use the full desktop area rather than the full screen.

**Comment by danwalmsley** [comment]
I'm also unsure how this can be unit tested? It doesn't really affect state of Window, only tell backend how to behave, unless we have any win32 specific unit tests?

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 1]
```diff
@@ -184,42 +186,53 @@ public void Hide()
...
             else
+            {
                 newRect = new Rect(
-                    windowRect.left + oldThickness.Left,
-                    windowRect.top + oldThickness.Top,
-                    (windowRect.right - windowRect.left) - (oldThickness.Left + oldThickness.Right),
-                    (windowRect.bottom - windowRect.top) - (oldThickness.Top + oldThickness.Bottom));
-            UnmanagedMethods.SetWindowPos(_hwnd, IntPtr.Zero, (int) newRect.X, (int) newRect.Y, (int) newRect.Width,
-                (int) newRect.Height,
+                windowRect.left + oldThickness.Left,
```
Comment (grokys) [inline src/Windows/Avalonia.Win32/WindowImpl.cs]: Indenting seems to be broken here.

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 2]
```diff
@@ -666,22 +679,57 @@ private void ShowWindow(WindowState state)
...
             switch (state)
             {
                 case WindowState.Minimized:
                     command = UnmanagedMethods.ShowWindowCommand.Minimize;
                     break;
                 case WindowState.Maximized:
                     command = UnmanagedMethods.ShowWindowCommand.Maximize;
+
+                    if (!_decorated && !_ignoreTaskBarWhenMaximized)
+                    {
```
Comment (grokys) [inline src/Windows/Avalonia.Win32/WindowImpl.cs]: Maybe the block below might be best refactored out into a separate method?

**Review by grokys — CHANGES_REQUESTED** [review]

File: src/Avalonia.Controls/Window.cs [hunk 3]
```diff
@@ -158,6 +168,16 @@ public bool HasSystemDecorations
         }
 
         /// <summary>
+        /// Sets if the Window should Covert the taskbar when Maximized. Only applies to Windows
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: Covert? As in covert operations? ;) Also nit: we don't capitalize "Window" elsewhere, unless we're doing a `<see cref="Window">`

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 4]
```diff
@@ -184,42 +186,52 @@ public void Hide()
...
             if (value)
             {
                 var thickness = BorderThickness;
+
                 newRect = new Rect(
-                    windowRect.left - thickness.Left,
-                    windowRect.top - thickness.Top,
-                    (windowRect.right - windowRect.left) + (thickness.Left + thickness.Right),
-                    (windowRect.bottom - windowRect.top) + (thickness.Top + thickness.Bottom));
+                windowRect.left - thickness.Left,
```
Comment (grokys) [inline src/Windows/Avalonia.Win32/WindowImpl.cs]: Indentation still appears to be wrong here.

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 5]
```diff
@@ -184,42 +186,52 @@ public void Hide()
...
                 newRect = new Rect(
-                    windowRect.left + oldThickness.Left,
-                    windowRect.top + oldThickness.Top,
-                    (windowRect.right - windowRect.left) - (oldThickness.Left + oldThickness.Right),
-                    (windowRect.bottom - windowRect.top) - (oldThickness.Top + oldThickness.Bottom));
-            UnmanagedMethods.SetWindowPos(_hwnd, IntPtr.Zero, (int) newRect.X, (int) newRect.Y, (int) newRect.Width,
-                (int) newRect.Height,
-                UnmanagedMethods.SetWindowPosFlags.SWP_NOZORDER | UnmanagedMethods.SetWindowPosFlags.SWP_NOACTIVATE);
+                windowRect.left + oldThickness.Left,
+                windowRect.top + oldThickness.Top,
```
Comment (grokys) [inline src/Windows/Avalonia.Win32/WindowImpl.cs]: And here.

File: src/Avalonia.Controls/Window.cs [hunk 6]
```diff
@@ -65,6 +65,12 @@ public class Window : TopLevel, IStyleable, IFocusScope, ILayoutRoot, INameScope
             AvaloniaProperty.Register<Window, bool>(nameof(HasSystemDecorations), true);
 
         /// <summary>
+        /// When system window decorations are disables sets if the Window when maximized ignores the taskbar.
```
Comment (grokys) [inline src/Avalonia.Controls/Window.cs]: Could we use "covers the taskbar" here as well?

**Review by grokys — CHANGES_REQUESTED** [review]

**Review by grokys — APPROVED** [review]

**Comment by grokys** [comment]
Yep, LGTM!

**Pull request merged**